nbs/utils.ipynb

```diff
@@ -547,6 +556,159 @@
+    "            One of the supported methods for the computation of conformal prediction:\n",
+    "            conformal_error or conformal_distribution\n",
+    "        \"\"\"\n",
+    "        if n_windows < 2:\n",
+    "            raise ValueError(\n",
+    "                \"You need at least two windows to compute conformal intervals\"\n",
+    "            )\n",
+    "        allowed_methods = [\"conformal_error\", \"conformal_distribution\"]\n",
+    "        if method not in allowed_methods:\n",
+    "            raise ValueError(f\"method must be one of {allowed_methods}\")\n",
+    "        self.n_windows = n_windows\n",
+    "        self.method = method\n",
+    "\n",
+    "    def __repr__(self):\n",
+    "        return f\"ConformalIntervals(n_windows={self.n_windows}, method='{self.method}')\""
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#| export\n",
+    "\n",
+    "def add_conformal_distribution_intervals(\n",
+    "    fcst_df: DataFrame, \n",
+    "    cs_df: DataFrame,\n",
+    "    model_names: List[str],\n",
+    "    level: List[Union[int, float]],\n",
+    "    cs_n_windows: int,\n",
+    "    n_series: int,\n",
+    "    horizon: int,\n",
+    ") -> DataFrame:\n",
+    "    \"\"\"\n",
+    "    Adds conformal intervals to a `fcst_df` based on conformal scores `cs_df`.\n",
+    "    `level` should be already sorted. This strategy creates forecasts paths\n",
+    "    based on errors and calculate quantiles using those paths.\n",
+    "    \"\"\"\n",
+    "    fcst_df = ufp.copy_if_pandas(fcst_df, deep=False)\n",
+    "    alphas = [100 - lv for lv in level]\n",
+    "    cuts = [alpha / 200 for alpha in reversed(alphas)]\n",
+    "    cuts.extend(1 - alpha / 200 for alpha in alphas)\n",
+    "    for model in model_names:\n",
+    "        scores = cs_df[model].to_numpy().reshape(cs_n_windows, n_series, horizon)\n",
+    "        # restrict scores to horizon\n",
+    "        scores = scores[:,:,:horizon]\n",
+    "        mean = fcst_df[model].to_numpy().reshape(1, n_series, -1)\n",
+    "        scores = np.vstack([mean - scores, mean + scores])\n",
+    "        quantiles = np.quantile(\n",
+    "            scores,\n",
+    "            cuts,\n",
+    "            axis=0,\n",
+    "        )\n",
+    "        quantiles = quantiles.reshape(len(cuts), -1).T\n",
+    "        lo_cols = [f\"{model}-conformal-lo-{lv}\" for lv in reversed(level)]\n",
+    "        hi_cols = [f\"{model}-conformal-hi-{lv}\" for lv in level]\n",
+    "        out_cols = lo_cols + hi_cols\n",
+    "        fcst_df = ufp.assign_columns(fcst_df, out_cols, quantiles)\n",
+    "    return fcst_df"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#| export\n",
+    "\n",
+    "def add_conformal_error_intervals(\n",
+    "    fcst_df: DataFrame, \n",
```

Comment:
@jose-moralez I think this should take `DFType` instead of `DataFrame`, similar as to how MLForecast does it, right? 

See [here](https://github.com/Nixtla/mlforecast/blob/fa402075a7764a1374438261dc20259dee6ca19c/mlforecast/forecast.py#L77)